File: web/src/notification-preferences.ts
export type NotificationEvent =
  | "approval"
  | "failure"
  | "completion"
  | "interruption"
  | "longRunning";

export interface RepositoryNotificationOverride {
  notifyApprovals?: boolean;
  notifyFailures?: boolean;
  notifyCompletions?: boolean;
  notifyInterruptions?: boolean;
  notifyLongRunning?: boolean;
}

export interface NotificationPreferences {
  notifyApprovals: boolean;
  notifyFailures: boolean;
  notifyCompletions: boolean;
  notifyInterruptions: boolean;
  notifyLongRunning: boolean;
  longRunningMinutes: number;
  quietHoursEnabled: boolean;
  quietStart: string;
  quietEnd: string;
  criticalBypassQuietHours: boolean;
  repositoryOverrides: Record<string, RepositoryNotificationOverride>;
}

export const DEFAULT_NOTIFICATION_PREFERENCES: NotificationPreferences = {
  notifyApprovals: true,
  notifyFailures: true,
  notifyCompletions: true,
  notifyInterruptions: false,
  notifyLongRunning: false,
  longRunningMinutes: 15,
  quietHoursEnabled: false,
  quietStart: "22:00",
  quietEnd: "07:00",
  criticalBypassQuietHours: false,
  repositoryOverrides: {},
};

const EVENT_KEYS: Record<NotificationEvent, keyof RepositoryNotificationOverride> = {
  approval: "notifyApprovals",
  failure: "notifyFailures",
  completion: "notifyCompletions",
  interruption: "notifyInterruptions",
  longRunning: "notifyLongRunning",
};

export function normalizeNotificationPreferences(value: unknown): NotificationPreferences {
  const candidate = value && typeof value === "object"
    ? value as Partial<NotificationPreferences>
    : {};
  const repositories = candidate.repositoryOverrides && typeof candidate.repositoryOverrides === "object"
    ? Object.fromEntries(Object.entries(candidate.repositoryOverrides).flatMap(([identity, override]) => {
      if (!identity || !override || typeof override !== "object") return [];
      const normalized = normalizeOverride(override as RepositoryNotificationOverride);
      return Object.keys(normalized).length ? [[identity, normalized]] : [];
    }).slice(-250))
    : {};
  return {
    notifyApprovals: booleanOr(candidate.notifyApprovals, true),
    notifyFailures: booleanOr(candidate.notifyFailures, true),
    notifyCompletions: booleanOr(candidate.notifyCompletions, true),
    notifyInterruptions: booleanOr(candidate.notifyInterruptions, false),
    notifyLongRunning: booleanOr(candidate.notifyLongRunning, false),
    longRunningMinutes: integerBetween(candidate.longRunningMinutes, 1, 1440, 15),
    quietHoursEnabled: booleanOr(candidate.quietHoursEnabled, false),
    quietStart: validTime(candidate.quietStart) ? candidate.quietStart : "22:00",
    quietEnd: validTime(candidate.quietEnd) ? candidate.quietEnd : "07:00",
    criticalBypassQuietHours: booleanOr(candidate.criticalBypassQuietHours, false),
    repositoryOverrides: repositories,
  };
}

export function effectiveNotificationPreferences(
  preferences: NotificationPreferences,
  repositoryIdentity: string,
): NotificationPreferences {
  const override = preferences.repositoryOverrides[repositoryIdentity] ?? {};
  return { ...preferences, ...override };
}

export function shouldNotify(
  event: NotificationEvent,
  preferences: NotificationPreferences,
  repositoryIdentity: string,
  at = new Date(),
): boolean {
  const effective = effectiveNotificationPreferences(preferences, repositoryIdentity);
  if (!effective[EVENT_KEYS[event]]) return false;
  if (!isQuietTime(effective, at)) return true;
  return effective.criticalBypassQuietHours && (event === "approval" || event === "failure");
}

export function isQuietTime(preferences: NotificationPreferences, at = new Date()): boolean {
  if (!preferences.quietHoursEnabled) return false;
  const start = timeMinutes(preferences.quietStart);
  const end = timeMinutes(preferences.quietEnd);
  if (start === null || end === null || start === end) return false;
  const current = at.getHours() * 60 + at.getMinutes();
  return start < end
    ? current >= start && current < end
    : current >= start || current < end;
}

export function setRepositoryOverride(
  preferences: NotificationPreferences,
  repositoryIdentity: string,
  update: RepositoryNotificationOverride,
): NotificationPreferences {
  const override = normalizeOverride({
    ...preferences.repositoryOverrides[repositoryIdentity],
    ...update,
  });
  const repositoryOverrides = { ...preferences.repositoryOverrides };
  if (Object.keys(override).length) repositoryOverrides[repositoryIdentity] = override;
  else delete repositoryOverrides[repositoryIdentity];
  return { ...preferences, repositoryOverrides };
}

function normalizeOverride(value: RepositoryNotificationOverride): RepositoryNotificationOverride {
  return Object.fromEntries(Object.entries(value).filter(([key, enabled]) =>
    Object.values(EVENT_KEYS).includes(key as keyof RepositoryNotificationOverride) && typeof enabled === "boolean"
  )) as RepositoryNotificationOverride;
}

function booleanOr(value: unknown, fallback: boolean): boolean {
  return typeof value === "boolean" ? value : fallback;
}

function integerBetween(value: unknown, minimum: number, maximum: number, fallback: number): number {
  return typeof value === "number" && Number.isInteger(value)
    ? Math.min(maximum, Math.max(minimum, value))
    : fallback;
}

function validTime(value: unknown): value is string {
  return typeof value === "string" && /^([01]\d|2[0-3]):[0-5]\d$/.test(value);
}

function timeMinutes(value: string): number | null {
  if (!validTime(value)) return null;
  const [hours, minutes] = value.split(":").map(Number);
  return hours * 60 + minutes;
}

File: android/app/src/main/java/net/kaltner/foreman/NotificationPreferences.kt
package net.kaltner.foreman

import android.content.Context
import java.util.Calendar
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

private val notificationPreferencesJson = Json { ignoreUnknownKeys = true; encodeDefaults = true }

internal fun encodeNotificationPreferences(value: NotificationPreferences): String =
    notificationPreferencesJson.encodeToString(value.normalized())

internal fun decodeNotificationPreferences(value: String?): NotificationPreferences =
    runCatching {
        notificationPreferencesJson.decodeFromString<NotificationPreferences>(value.orEmpty())
    }.getOrDefault(NotificationPreferences()).normalized()

internal enum class NotificationEvent {
    Approval,
    Failure,
    Completion,
    Interruption,
    LongRunning,
}

@Serializable
data class RepositoryNotificationOverride(
    val notifyApprovals: Boolean? = null,
    val notifyFailures: Boolean? = null,
    val notifyCompletions: Boolean? = null,
    val notifyInterruptions: Boolean? = null,
    val notifyLongRunning: Boolean? = null,
)

@Serializable
data class NotificationPreferences(
    val notifyApprovals: Boolean = true,
    val notifyFailures: Boolean = true,
    val notifyCompletions: Boolean = true,
    val notifyInterruptions: Boolean = false,
    val notifyLongRunning: Boolean = false,
    val longRunningMinutes: Int = 15,
    val quietHoursEnabled: Boolean = false,
    val quietStart: String = "22:00",
    val quietEnd: String = "07:00",
    val criticalBypassQuietHours: Boolean = false,
    val repositoryOverrides: Map<String, RepositoryNotificationOverride> = emptyMap(),
)

internal fun NotificationPreferences.normalized(): NotificationPreferences =
    copy(
        longRunningMinutes = longRunningMinutes.coerceIn(1, 1_440),
        quietStart = quietStart.takeIf(::validClockTime) ?: "22:00",
        quietEnd = quietEnd.takeIf(::validClockTime) ?: "07:00",
        repositoryOverrides = repositoryOverrides.filterKeys(String::isNotBlank).entries.toList().takeLast(250)
            .associate { it.key to it.value },
    )

internal fun NotificationPreferences.eventEnabled(
    event: NotificationEvent,
    repositoryIdentity: String,
): Boolean {
    val override = repositoryOverrides[repositoryIdentity]
    return when (event) {
        NotificationEvent.Approval -> override?.notifyApprovals ?: notifyApprovals
        NotificationEvent.Failure -> override?.notifyFailures ?: notifyFailures
        NotificationEvent.Completion -> override?.notifyCompletions ?: notifyCompletions
        NotificationEvent.Interruption -> override?.notifyInterruptions ?: notifyInterruptions
        NotificationEvent.LongRunning -> override?.notifyLongRunning ?: notifyLongRunning
    }
}

internal fun NotificationPreferences.shouldNotify(
    event: NotificationEvent,
    repositoryIdentity: String,
    nowMillis: Long = System.currentTimeMillis(),
): Boolean {
    if (!eventEnabled(event, repositoryIdentity)) return false
    if (!isQuietTime(nowMillis)) return true
    return criticalBypassQuietHours &&
        (event == NotificationEvent.Approval || event == NotificationEvent.Failure)
}

internal fun NotificationPreferences.isQuietTime(nowMillis: Long = System.currentTimeMillis()): Boolean {
    if (!quietHoursEnabled) return false
    val start = clockMinutes(quietStart) ?: return false
    val end = clockMinutes(quietEnd) ?: return false
    if (start == end) return false
    val calendar = Calendar.getInstance().apply { timeInMillis = nowMillis }
    val current = calendar.get(Calendar.HOUR_OF_DAY) * 60 + calendar.get(Calendar.MINUTE)
    return if (start < end) current in start until end else current >= start || current < end
}

internal class NotificationPreferenceStore(context: Context) {
    private val preferences =
        context.getSharedPreferences(PREFERENCE_FILE, Context.MODE_PRIVATE)

    fun load(hostId: String? = null): NotificationPreferences =
        hostId?.let(::loadHostOverride) ?: loadGlobal()

    fun loadGlobal(): NotificationPreferences = decode(preferences.getString(GLOBAL_KEY, null))

    fun loadHostOverride(hostId: String): NotificationPreferences? =
        preferences.getString(hostKey(hostId), null)?.let(::decode)

    fun hasHostOverride(hostId: String?): Boolean =
        hostId != null && preferences.contains(hostKey(hostId))

    fun save(value: NotificationPreferences, hostId: String? = null) {
        preferences.edit()
            .putString(hostId?.let(::hostKey) ?: GLOBAL_KEY, encode(value))
            .apply()
    }

    fun clearHostOverride(hostId: String) {
        preferences.edit().remove(hostKey(hostId)).apply()
    }

    internal fun encode(value: NotificationPreferences): String =
        encodeNotificationPreferences(value)

    internal fun decode(value: String?): NotificationPreferences =
        decodeNotificationPreferences(value)

    private fun hostKey(hostId: String) = "host.$hostId"

    companion object {
        private const val PREFERENCE_FILE = "foreman_notification_preferences"
        private const val GLOBAL_KEY = "global.v1"
    }
}

private fun validClockTime(value: String): Boolean =
    Regex("^([01]\\d|2[0-3]):[0-5]\\d$").matches(value)

private fun clockMinutes(value: String): Int? {
    if (!validClockTime(value)) return null
    val (hours, minutes) = value.split(':').map(String::toInt)
    return hours * 60 + minutes
}
